### PR TITLE
[ios] Lock refresh rate to 80fps when threads are merged

### PR DESCRIPTION
**Pull request by cyanglaz**
When there are PlatformViews on the screen (threads merged), Flutter engine cannot consistently hit 120 fps, especially when the PlatformView is animating (scrolling etc), thus producing janks. 
This PR locks the refresh rate to 80fps to avoid janks.

Fixes https://github.com/flutter/flutter/issues/116640

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm
Comment (cyanglaz) [inline shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm:46]: I originally set it to 60, but it looks like we can def hit 80 consistently. I decided to lock it at 80 so the scrolling with PlatformView is still smoother than 60fps.

This might change when we re-evaluate what the default fps we want the Flutter apps to run on after we implement a new API for the APP to set refresh rates, see https://github.com/flutter/flutter/issues/119268

**Comment by chinmaygarde** [comment]
cc @dnfield. Instead of hardcoding values, perhaps we should determine a frame rate we can hit consistently and then stick to it for a certain period of time before re-evaluating?

**Comment by cyanglaz** [comment]
> perhaps we should determine a frame rate we can hit consistently and then stick to it for a certain period of time before re-evaluating?

I have tried this but I wasn't satisfied with the out come with various ways to determine the refresh rates.
Because PlatformView scrolling are mostly able to hit 120fps with only a few random bad frames, it is hard to determine when to update the refresh rate to 80fps.

I think the best way to go is to have a dart:ui API for the app or the framework set the refresh rate for certain situations: https://github.com/flutter/flutter/issues/119268

**Comment by cyanglaz** [comment]
The one drawback: 
If there is a PlatformView in the scene and a high-impact animation without PlatformView happening (a non-platformview widget only scroll view, or a non-platformview widget page transitioning into the scene), with the refresh rate set to 80fps, the animation is not able to run 120fps as it could have.

**Comment by delfme** [comment]
@cyanglaz does this issue need a CP fix? I lately noticed that if an android platform view is on the scene, that impairs scrolling of flutter listView. Although it seems that scrolling is smooth if I keep the finger on the screen while performing the action, so maybe gesture is also involved here. https://github.com/flutter/flutter/issues/119337

**Comment by cyanglaz** [comment]
> @cyanglaz does this issue need a CP fix? I lately noticed that if an android platform view is on the scene, that impairs scrolling of flutter listView. Although it seems that scrolling is smooth if I keep the finger on the screen while performing the action, so maybe gesture is also involved here. [flutter/flutter#119337](https://github.com/flutter/flutter/issues/119337)

This is not related to Android. On iOS, we did have a fix for gesture https://github.com/flutter/engine/commit/2c62210af44568cb3add38ab6f7f77424c27e498
I don't have a strong reason to push this to a CP. Because we don't have a reliable way to test refresh rates, I would rather have this baked on main for a while to gather feedbacks.

File: shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm
Comment (jmagman) [reply shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm:46]: Can you add a comment for why you chose this?  What hardware did you try?

File: shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm:18]: Can you add a comment for what this is?  4 what?

File: shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm:146]: Might be better to move this method to line 79 so the diff history from `setMaxRefreshRateIfEnabled` isn't totally lost.

File: shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm:188]:
```suggestion
+ (BOOL)minimumFrameDurationDisabled {
```

File: shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.h
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.h:18]: Can you add a comment here?  What does it mean to disable this?  Also might be easier to understand as a positive: `minimumFrameDurationEnabled`

```suggestion
@property(class, nonatomic, readonly) BOOL minimumFrameDurationDisabled;
```

File: shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.h
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.h:18]: Pro-motion iPad Pros too, right?

File: shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm
Comment (jmagman) [inline shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm:19]: Why was it 4 before, out of curiosity?

Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.mm:19]: For some reason I thought the least difference between each available FPS was 5, but I double checked the list here: https://developer.apple.com/documentation/quartzcore/optimizing_promotion_refresh_rates_for_iphone_13_pro_and_ipad_pro?language=objc and found there are refresh rates that closer. Now I just saw 15 and 16 both in the list and I think I will adjust it smaller, probably 0.1 to be safe.

File: shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.h
Comment (cyanglaz) [reply shell/platform/darwin/ios/framework/Source/vsync_waiter_ios.h:18]: This is only for iPhone, iPad always has high refresh rate on, let me add a comment about it too.

**Review by jmagman — COMMENTED** [review]
This overall looks like an okay compromise to avoid jank until https://github.com/flutter/flutter/issues/119268 is done.  However it would get too get a LGTM from someone more familiar with the vsync waiter code (@iskakaushik?)

**Review by jmagman — APPROVED** [review]

**Comment by zanderso** [comment]
Note from PR review triage: This looks like it will need to be rebased.

**Review by iskakaushik — APPROVED** [review]
LGTM

**Comment by cyanglaz** [comment]
https://github.com/flutter/engine/pull/39359#pull claims to reduce janks. I would like to test it before landing this PR. This PR might not be necessary if 39359 fixed the jank issue for PlatformView as well.

**Comment by cyanglaz** [comment]
https://github.com/flutter/engine/pull/39359#pull doesn't fix the issue, we will have to land this.

**Comment by delfme** [comment]
@cyanglaz I see this fixes jittery in case of a platform view being animated on the scene, but I still see random janks when animating a native texfield. Testing on iphone 13 pro ios 16.1.1


https://user-images.githubusercontent.com/53510751/217750858-55d6a688-f766-4688-ae3b-e7b9a7d97a81.MOV

**Comment by cyanglaz** [comment]
@delfme Thanks for reporting this. 
The textfield Janks is likely a separate issue. Do you mind open a new github issue, include details, with a reduced sample code to repro?

**Comment by delfme** [comment]
@cyanglaz I opened it for android today. Will better investigate issue on iOS and lyk.

**Pull request merged**
